src/Time/Actions/LimitTimeStepToExpirationTimes.hpp

```diff
@@ -0,0 +1,175 @@
+    // First, check whether the time stepper uses substeps
+    const TimeStepper& time_stepper = db::get<Tags::TimeStepper<>>(box);
+    if (time_stepper.number_of_substeps() > 0) {
+      // Next, check if the current time step is at the beginning of a full step
+      const auto& time_step_id = db::get<Tags::TimeStepId>(box);
+      if (time_step_id.substep() == 0) {
+        // Get the functions of time and, if not empty, loop over them,
+        // finding the minimum expiration time
+        const auto& functions_of_time =
+            get<domain::Tags::FunctionsOfTime>(cache);
+        if (not functions_of_time.empty()) {
+          double min_expiration_time{std::numeric_limits<double>::max()};
+          for (const auto& [name, f_of_t] : functions_of_time) {
+            if (f_of_t->time_bounds()[1] < min_expiration_time) {
+              min_expiration_time = f_of_t->time_bounds()[1];
+            }
+          }
+
+          ASSERT(db::get<Tags::Time>(box) <= min_expiration_time,
+                 "Time (" << std::setprecision(20) << db::get<Tags::Time>(box)
+                          << ") is greater than expiration time ("
+                          << min_expiration_time << ") by an amount of "
+                          << db::get<Tags::Time>(box) - min_expiration_time);
+          if (db::get<Tags::Time>(box) == min_expiration_time) {
+            return {Parallel::AlgorithmExecution::Continue, std::nullopt};
+          }
+
+          const auto& initial_time_step{db::get<Tags::TimeStep>(box)};
+          const double step_end =
+              (time_step_id.step_time() + initial_time_step).value();
+          if (time_step_id.step_time().value() +
+                  2.0 * initial_time_step.value() <=
+              min_expiration_time) {
+            // The expiration times are far in the future (more than
+            // two steps).  No need to adjust anything.
+            return {Parallel::AlgorithmExecution::Continue, std::nullopt};
+          }
+
+          const double start{initial_time_step.slab().start().value()};
+          Slab new_slab{};
+          if (step_end + slab_rounding_error(time_step_id.step_time()) >=
+              min_expiration_time) {
+            // The minimum expiration time is within the next step (or
+            // within roundoff).  Step to it.
+            new_slab = Slab(start, min_expiration_time);
+          } else {
+            // The minimum expiration time is not within the next
+            // step, but is within two steps.  Shrink the step size to
+            // avoid taking a tiny step after this one.
+
+            // new_slab = initial_time_step.slab().with_duration_to_end(
+            //     0.51 * (min_expiration_time - start));
+            // Note: 0.51 instead of 0.5 to guarantee no roundoff affecting
+            // what happens on the next step.
+            new_slab = Slab(start, 0.51 * min_expiration_time + 0.49 * start);
```

Comment:
Weighting the average towards `min_expiration_time` like this is good, but add a check to avoid increasing the step size from its current value.